cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,254 @@
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "default.user-data",
+    ]
+
+
+DEFAULT_INSTANCE_ID = "datasource-wsl"
+
+
+def load_metadata_iid(cloudinitdir: PurePath, instance_name: str) -> str:
+    """
+    Returns the relevant metadata loaded from cloudinit dir based on the
+    instance name
+    """
+    raw = dict()
+    try:
+        raw = json.loads(
+            util.load_binary_file(
+                os.path.join(
+                    cloudinitdir.as_posix(), "%s.meta-data" % instance_name
+                )
+            )
+        )
+
+    except IOError as err:
+        LOG.debug(
+            "Failed to load metadata file from %s for instance %s: %s",
+            cloudinitdir.as_posix(),
+            instance_name,
+            err,
+        )
+
+    return raw.get("instance-id", DEFAULT_INSTANCE_ID)
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def find_user_data_file(self, seed_dir: PurePath) -> PurePath:
+        """
+        Finds the most precendent of the candidate files that may contain
+        user-data, if any, or None otherwise.
+        """
+
+        # Notice that by default file name casing is irrelevant here. Windows
+        # filenames are case insensitive. Even though accessed through Linux,
+        # path translation just works with whichever casing we try.
+        # But users can change that behavior with configuration
+        # (ref https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity),
+        # thus  better prevent it by always relying on case insensitive match.
+        existing_files = {
+            ef.name.casefold(): ef.path for ef in os.scandir(seed_dir)
+        }
+        if not existing_files:
+            raise IOError("%s directory is empty" % seed_dir)
+
+        folded_names = [
+            f.casefold()
+            for f in candidate_user_data_file_names(self.instance_name)
+        ]
+        for filename in folded_names:
+            if filename in existing_files.keys():
+                return PurePath(existing_files[filename])
+
+        raise IOError(
+            "%s doesn't contain any of the expected user-data files" % seed_dir
+        )
+
+    def check_instance_id(self, sys_cfg):
+        current = self.get_instance_id()
+        if not current:
+            return None
+
+        try:
+            return current == load_metadata_iid(
+                cloud_init_data_dir(), self.instance_name
+            )
+        except IOError as err:
+            LOG.error("Could not load updated instance ID: %s", err)
+            return None
+
+    def _get_data(self) -> bool:
+        self.vendordata_raw = None
+        self.metadata = dict()
+        seed_dir = cloud_init_data_dir()
+        self.instance_name = instance_name()
+
+        try:
+            self.metadata["instance-id"] = load_metadata_iid(
+                seed_dir, self.instance_name
+            )
```

Comment:
I was hesitant to allow all sorts of metadata relying on the base class behavior and figuring out some of that is not applicable under the WSL environment, so instead of blocklisting the unsuitable bits I'd though we should just allow-list the parts we know for sure to work. But I think you convinced me the proper way to deal with this situation  is by figuring out the usage in the wild and adapting the unwanted behavior as we find them.